webapps-backup/src/main/java/io/camunda/webapps/backup/repository/opensearch/OpensearchBackupRepository.java

```diff
@@ -0,0 +1,475 @@
+      LOGGER.info("Snapshot done: {}", snapshotInfo.uuid());
+      onSuccess.run();
+    } else if (FAILED.equals(SnapshotState.valueOf(snapshotInfo.state()))) {
+      LOGGER.error(
+          "Snapshot taking failed for {}, reason {}", snapshotInfo.uuid(), snapshotInfo.reason());
+      // No need to continue
+      onFailure.run();
+    } else {
+      LOGGER.warn(
+          "Snapshot state is {} for snapshot {}", snapshotInfo.state(), snapshotInfo.uuid());
+      onSuccess.run();
+    }
+  }
+
+  private List<OpenSearchSnapshotInfo> findSnapshots(
+      final String repositoryName, final Long backupId) {
+    final var requestBuilder =
+        getSnapshotRequestBuilder(
+                repositoryName, snapshotNameProvider.getSnapshotNamePrefix(backupId) + "*")
+            .build();
+
+    final OpenSearchGetSnapshotResponse response;
+    try {
+      response =
+          OpenSearchGetSnapshotResponse.fromResponse(
+              openSearchClient.snapshot().get(requestBuilder));
+      return response.snapshots();
+    } catch (final Exception e) {
+      if (isSnapshotMissingException(e)) {
+        // no snapshot with given backupID exists
+        throw new ResourceNotFoundException(format("No backup with id [%s] found.", backupId));
+      }
+      if (isRepositoryMissingException(e)) {
+        final String reason = noRepositoryErrorMessage(repositoryName);
+        throw new GenericBackupException(reason);
+      }
+      final String reason =
+          format("Exception occurred when searching for backup with ID [%s].", backupId);
+      throw new GenericBackupException(reason, e);
+    }
+  }
+
+  private BackupStateDto getState(
+      final List<OpenSearchSnapshotInfo> snapshots, final Integer expectedSnapshotsCount) {
+    if (snapshots.size() == expectedSnapshotsCount
+        && snapshots.stream().map(OpenSearchSnapshotInfo::getState).allMatch(SUCCESS::equals)) {
+      return BackupStateDto.COMPLETED;
+    } else if (snapshots.stream()
+        .map(OpenSearchSnapshotInfo::getState)
+        .anyMatch(s -> FAILED.equals(s) || PARTIAL.equals(s))) {
+      return BackupStateDto.FAILED;
+    } else if (snapshots.stream().map(OpenSearchSnapshotInfo::getState).anyMatch(STARTED::equals)) {
+      return BackupStateDto.IN_PROGRESS;
+    } else if (snapshots.size() < expectedSnapshotsCount) {
+      if (isIncompleteCheckTimedOut(
+          backupProps.incompleteCheckTimeoutInSeconds(),
+          snapshots.getLast().getEndTimeInMillis())) {
+        return BackupStateDto.INCOMPLETE;
+      } else {
+        return BackupStateDto.IN_PROGRESS;
+      }
+    } else {
+      return BackupStateDto.FAILED;
+    }
+  }
+
+  private GetBackupStateResponseDto toGetBackupStateResponseDto(
+      final Long backupId, final List<OpenSearchSnapshotInfo> snapshots) {
+    final GetBackupStateResponseDto response = new GetBackupStateResponseDto(backupId);
+    final Metadata metadata =
+        Metadata.fromMetadata(
```

Comment:
Deserializes Metadata from `Map<String,JsonData>` instead of using ObjectMapper, see above